src/canard.c

```diff
@@ -1,6 +1,845 @@
+                            uint8_t destination_node_id,
+                            uint64_t data_type_signature,
+                            uint16_t data_type_id,
+                            uint8_t* inout_transfer_id,
+                            uint8_t priority,
+                            CanardRequestResponse kind,
+                            const void* payload,
+                            uint16_t payload_len)
+{
+  if (payload == NULL)
+  {
+      return -1;
+  }
+  if (canardGetLocalNodeID(ins) == 0)
+  {
+    return -1;
+  }
+  if (priority > 31)
+  {
+      return -1;
+  }
+
+  const uint32_t can_id = ((uint32_t)priority << 24) | ((uint32_t)data_type_id << 16) | 
+                            ((uint32_t)kind << 15) | ((uint32_t)destination_node_id << 8) | 
+                              (1 << 7) | (uint32_t)canardGetLocalNodeID(ins);
+  uint16_t crc = 0xFFFFU;
+
+  if (payload_len > 7)
+  {
+    crc = crc_add_signature(crc, data_type_signature);
+    crc = crc_add(crc, payload, payload_len);
+  }
+
+  canardEnqueueData(ins, can_id, inout_transfer_id, crc, payload, payload_len);
+
+  tidIncrement(inout_transfer_id);
+
+  return 1;
+}
+
+/**
+ * Returns a pointer to the top priority frame in the TX queue.
+ * Returns NULL if the TX queue is empty.
+ */
+const CanardCANFrame* canardPeekTxQueue(const CanardInstance* ins)
+{
+  if (ins->tx_queue == NULL)
+  {
+    return NULL;
+  }
+  return &ins->tx_queue->frame;
+}
+
+/**
+ * Removes the top priority frame from the TX queue.
+ */
+void canardPopTxQueue(CanardInstance* ins) 
+{
+  CanardTxQueueItem* item = ins->tx_queue;
+  ins->tx_queue = item->next;
+  canardFreeBlock(&ins->allocator, item);
+}
+
+/**
+ * Processes a received CAN frame with a timestamp.
+ */
+void canardHandleRxFrame(CanardInstance* ins, const CanardCANFrame* frame, uint64_t timestamp_usec)
+{
+  
+  uint8_t transfer_type = canardTransferType(frame->id);
+  uint8_t priority = CANARD_PRIORITY_FROM_ID(frame->id);
+  uint8_t source_node_id = CANARD_SOURCE_ID_FROM_ID(frame->id);
+  uint8_t destination_node_id = (transfer_type == CanardTransferTypeBroadcast) ? 0 : CANARD_DEST_ID_FROM_ID(frame->id);
+  uint16_t data_type_id = canardDataType(frame->id);
+  uint32_t transfer_descriptor = CANARD_MAKE_TRANSFER_DESCRIPTOR(data_type_id, transfer_type, source_node_id, destination_node_id);
+  
+  CanardRxState* rxstate;
+  unsigned char tail_byte = frame->data[frame->data_len-1];
+
+  if (IS_START_OF_TRANSFER(tail_byte))
+  {
+    uint64_t data_type_signature;
+
+    if(ins->should_accept(ins, &data_type_signature, data_type_id, transfer_type, source_node_id))
+    {
+      rxstate = canardRxStateTraversal(ins, transfer_descriptor);
+      rxstate->calculated_crc = crc_add_signature(0xFFFFU, data_type_signature);
+    }
+  } else if (!IS_START_OF_TRANSFER(tail_byte))
+  {
+    rxstate = canardFindRxState(ins->rx_states, transfer_descriptor);
+    if (rxstate == NULL) {
+      return;
+    }
```

Comment:
I spy a bug here. If this is start of a transfer, and the application returns `false` from `should_accept()`, `rxstate` will be used uninitialized.

This condition should be placed after the closing brace below:

```c++
if (rxstate == NULL) {
    return;
}
```